socs/agents/hwp_gripper/agent.py

```diff
@@ -470,6 +525,87 @@ def run_and_append(func, *args, **kwargs):
+        """
+
+        self._ungrip_hwp(session, check_shutdown=True)
+        return True, "Finished ungripping procedure"
+
+    def _ungrip_hwp(self, session, check_shutdown=True):
+        """
+        Helper function to ungrip the HWP that can be called by the ungrip_hwp
+        task or by the shutdown procedure.  This will return grippers to their
+        home position, and retract as much as possible.
+
+        Args
+        ------
+        session (OpSession):
+            Session object for current operation. This function will add to
+            session data.
+
+        Notes:
+            The most recent data collected is stored in session data in the
+            structure:
+
+                >>> response.session['responses']
+                [{'result': True, 'log': [.., .., etc]},
+                                    ..
+                 {'result': True, 'log': [.., .., etc]}]
+        """
+        data = {
+            'responses': [],
+        }
+        session.data = data
+
+        def run_and_append(func, *args, **kwargs):
+            return_dict = self._run_client_func(func, *args, **kwargs)
+            data['responses'].append(return_dict)
+            return return_dict
+
+        run_and_append(self.client.reset, job='ungrip', check_shutdown=check_shutdown)
+        # Enable power to actuators
+        run_and_append(self.client.power, True, job='ungrip', check_shutdown=check_shutdown)
+        # Disable breaks
+        run_and_append(self.client.brake, False, job='ungrip', check_shutdown=check_shutdown)
+        # Ignore limit switches
+        run_and_append(self.client.force, True, job='ungrip', check_shutdown=check_shutdown)
+
+        # Send grippers to their home position
+        run_and_append(self.client.home, job='ungrip', check_shutdown=check_shutdown)
+
+        # Move actuator outwards as much as possible
+        for actuator in range(1, 4):
+            run_and_append(self.client.move, 'POS', actuator, -1.9,
+                           job='ungrip', check_shutdown=check_shutdown)
+        time.sleep(5)
+
+        # Enable brake
+        run_and_append(self.client.brake, True, job='ungrip', check_shutdown=check_shutdown)
+        # Power off actuators
+        run_and_append(self.client.power, False, job='ungrip', check_shutdown=check_shutdown)
+        # Enable limit switches
+        run_and_append(self.client.force, False, job='ungrip', check_shutdown=check_shutdown)
+        time.sleep(1)
+
+        # check limit switch state
+        self.monitor_state(session)
+        limit_switch_state = (not session['data']['state']['act1_limit_warm_grip_state']) & \
+                             (not session['data']['state']['act2_limit_warm_grip_state']) & \
+                             (not session['data']['state']['act3_limit_warm_grip_state'])
+
+        # We should stop schedule if we have an error in this task
+        assert limit_switch_state, "Failed to ungrip HWP."
```

Comment:
I'd like to see this check moved up into `ungrip()`, and rather than raising an `AssertionError` for force a task failure, do a `return False, "Failed to ungrip HWP"`, which cleanly indicates Task failure.